02cpp1/sec03Pointers.md

```diff
@@ -0,0 +1,364 @@
+---
+title: Pointers in C++
+---
+
+Estimated Reading Time: 1 hour
+
+# Pointers
+
+You'll already have used references to refer to objects without copying them. This is very useful for passing objects to functions without copy overheads, or to functions which also change those objects in ways that we want to persist once we've left that function's scope. There are, however, times where we cannot use references and must use pointers. Some limitations of references are:
+* References cannot be reassigned to refer to a new place in memory, they can only be assigned at initialisation. 
+* You cannot have references of references.
+* You cannot create circular references e.g. an object A which has a reference to object B, which in turn has a reference to object A. This would require reassignment of references to construct. 
+* References cannot be null. 
+* You cannot store references in container types like `vector`, `array`, `set`, or `map`. 
+
+In these cases, we use *pointers*. A pointer is variable which stores an address in memory where an object's data is located (we say that it "points to" this object), or the special value `nullptr`. Pointers give us much more flexibility than references, especially when writing classes for objects that need to point to other data (either of the same class, like in graph representation where nodes point to other nodes, or of another class). In modern C++ (since C++11) we usually declare a pointer using a *smart pointer*, of which there are three different kinds: unique pointers, shared pointers, and weak pointers. 
+
+## Optional Background: The Stack, the Heap, and Variable Scope
+
+We will go into more detail on memory structures later on in the course when we discuss performance programming. It can however be easier to understand the usage of pointers in C++ if we understand the difference between two different kinds of memory: the _stack_ and the _heap_. 
+- **Stack** memory is used to store all local variables, as well as the "call stack". The call stack contains information about the currently active functions, including the value of variables in each scope, and allows us to continue execution at the correct place in the program when we leave a function. Stack memory is fast, but limited in size. The amount of stack memory available is not known to the program at compile time, as stack memory is reserved for the program at runtime by the operating system. Using too much stack memory causes a _stack overflow_ error, which will cause your program to crash. When variables on the stack go out of scope then their destructor is called and their memory is freed. 
```

Comment:
"Stack memory is fast, but limited in size" - fast compared to what? Faster than the heap?